libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -0,0 +1,271 @@
+                CLIENT_ID to clientId,
+                REDIRECT_URI to redirectUri,
+                CODE_VERIFIER to codeVerifier,
+            )
+            authResponse.consumeQuietly()
+
+            val tokenRequest = RestRequest(
+                RestRequest.RestMethod.POST,
+                RestRequest.RestEndpoint.LOGIN,
+                authEndpoint + OAUTH_TOKEN_PATH,  // Full path for unauthenticated request
+                tokenRequestBody,
+                 null, // additionalHttpHeaders
+            )
+
+            // Second REST Call - token request with code verifier
+            val tokenResponse = suspendedRestCall(tokenRequest) ?: return NativeLoginResult.UnknownError
+            return if (tokenResponse.isSuccess) {
+                // Returns NativeLoginResult.Success or UnknownError based on if the user is created.
+                suspendFinishAuthFlow(tokenResponse)
+            } else {
+                NativeLoginResult.UnknownError
+            }
+        } else {
+            SalesforceSDKLogger.e(TAG, "Native Login Authorization Error: $authResponse")
+            return NativeLoginResult.InvalidCredentials
+        }
+    }
+
+    override fun getFallbackWebAuthenticationIntent(): Intent {
+        val context = SalesforceSDKManager.getInstance().appContext
+        val intent = Intent(context, SalesforceSDKManager.getInstance().webviewLoginActivityClass)
+        intent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
+        val options = SalesforceSDKManager.getInstance().loginOptions.asBundle()
+        options.putBoolean(BiometricAuthenticationManager.SHOW_BIOMETRIC, bioAuthLocked)
+        intent.putExtras(options)
+        Bundle().putParcelable(AccountManager.KEY_INTENT, intent)
+
+        return intent
+    }
+
+    @VisibleForTesting
+    internal fun isValidUsername(username: String): Boolean {
+        if (username.length > MAX_USERNAME_LENGTH) {
+            return false
+        }
+
+        return Regex(USERNAME_REGEX_PATTERN).matches(username)
+    }
+
+    @VisibleForTesting
+    internal fun isValidPassword(password: String): Boolean {
+        val containsNumber = password.contains("[0-9]".toRegex())
+        val containsLetter = password.contains("[A-Za-z]".toRegex())
+
+        return containsNumber && containsLetter && password.length >= MIN_PASSWORD_LENGTH
+                && password.toByteArray().size <= MAX_PASSWORD_LENGTH_BYTES
+    }
+
+    private suspend fun suspendFinishAuthFlow(tokenResponse: RestResponse): NativeLoginResult {
+        val appContext = SalesforceSDKManager.getInstance().appContext
+        val loginOptions = LoginOptions(loginUrl, redirectUri, clientId, emptyArray<String>())
+        val tokenEndpointResponse = OAuth2.TokenEndpointResponse(tokenResponse.rawResponse)
+        tokenResponse.consumeQuietly()
+
+        return suspendCoroutine { continuation ->
+            OAuthWebviewHelper(appContext, object : OAuthWebviewHelper.OAuthWebviewHelperEvents {
```

Comment:
I'm sure this is food for though much later, but it seems unwieldy to have to use `OAuthWebviewHelper` in this context.  I see how it works in the context of the codebase it is today, so no action required.  I have a work item in the backlog where it could be neat to see how we could separate `OAuthWebviewHelper` into smaller units of decoupled logic and then this could use just the one it needs without the two skipped implementations.

Just something to ponder for the future.

Reply:
Yeah, I agree.  `OAuthWebviewHelper` is designed around the webview lifecycle and the hooks into it are engrained all over.  I knew it would be too much work for this story/PR to try to separate it into smaller pieces but it would be really cool if we could do that in the future.